### PR TITLE
TYP: construction for `DatetimeArray`

### PR DESCRIPTION
**Pull request by cmp0xff**
Towards #1469.

- Fixed inheritance of several classes
- Added typing for `DatetimeArray`
- Added tests for `pd.array` that gives `DatetimeArray`

### CODE REVIEW

File: tests/arrays/test_datetime_array.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        )
+        assert_type(
+            pd.array([datetime(2052, 1, 5), pd.Timestamp(2, 1, 6)]), DatetimeArray
+        )
+
+        assert_type(pd.array([datetime(2061, 1, 5, 1), None]), DatetimeArray)
+        assert_type(pd.array([pd.Timestamp(1902, 1, 5, 3), None]), DatetimeArray)
+
+        assert_type(pd.array([datetime(1921, 1, 5, 1), pd.NaT]), DatetimeArray)  # type: ignore[assert-type]
+        assert_type(pd.array([pd.Timestamp(1872, 1, 5, 3), pd.NaT]), DatetimeArray)  # type: ignore[assert-type]
```
Comment (cmp0xff) [inline tests/arrays/test_datetime_array.py:86]: Mypy did not recognise these two. I've already removed `SequenceNotStr` in `construction.pyi` in this PR and fixed some errors, but I don't know how to fix further.

File: pandas-stubs/core/arrays/datetimes.pyi [hunk 2]
```diff
@@ -64,34 +63,55 @@ class DatetimeArray(DatetimeLikeArrayMixin, TimelikeOps, DatelikeOps):
...
-    is_leap_year = ...
+    def day_name(self, locale: str | None = None) -> np_1darray_object: ...
+    @property
+    def time(self) -> np_1darray_object: ...
+    @property
+    def timetz(self) -> np_1darray_object: ...
+    @property
+    def date(self) -> np_1darray_object: ...
+    @property
+    def year(self) -> np_1darray_int32: ...
```
Comment (loicdiridollou) [inline pandas-stubs/core/arrays/datetimes.pyi:76]: I think they may be int64 here (see in #1621)

File: tests/arrays/test_datetime_array.py (hunk 1 above)
Comment (loicdiridollou) [reply tests/arrays/test_datetime_array.py:86]: Do we have the correct overloads in pandas.core.construction? I don't see any that goes to DatetimeArray, maybe we need one like:
```python
@overload
 def array(  # type: ignore[overload-overlap] # pyright: ignore[reportOverlappingOverload]
    data: Sequence[pd.Timestamp | NAType | NaTType | None],
    dtype: DtypeArg | None = None,
    copy: bool = True,
) -> DatetimeArray: ...
```

Comment (cmp0xff) [reply tests/arrays/test_datetime_array.py:86]: At the moment we have
```py
@overload
def array(
    data: (
        tuple[_NaTDatetimeElement, ...]
        | MutableSequence[_NaTDatetimeElement]
        | np_ndarray
        | DatetimeArray
    ),
    dtype: (
        DatetimeTZDtype
        | PandasTimestampDtypeArg
        | np.dtype[np.datetime64]
        | NumpyTimestampDtypeArg
    ),
    copy: bool = True,
) -> DatetimeArray: ...
@overload
def array(  # type: ignore[overload-overlap]
    data: (  # TODO: merge the two Sequence's after 3.0 pandas-dev/pandas#57064
        Sequence[datetime | NaTType | None]
        | Sequence[np.datetime64 | NaTType | None]
        | np_ndarray_dt
        | DatetimeArray
    ),
    dtype: None = None,
    copy: bool = True,
) -> DatetimeArray: ...
```

File: pandas-stubs/core/arrays/datetimes.pyi (hunk 2 above)
Comment (cmp0xff) [reply pandas-stubs/core/arrays/datetimes.pyi:76]: Tests added 864007eeab2af9e2e764b6585c6f6a80d8bb6371

File: tests/arrays/test_datetime_array.py (hunk 1 above)
Comment (loicdiridollou) [reply tests/arrays/test_datetime_array.py:86]: I think once you have a pd.NA and a pd.NaT pandas at constrcution will move the pd.NA into pd.NaT so the bottom one we may need to have the pd.NA added to it

Comment (loicdiridollou) [reply tests/arrays/test_datetime_array.py:86]: I think it is an edge case anyway so not sure that should be a good practice to mix NA and NaT together.
